flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/memory/MemoryTierProducerAgent.java

```diff
@@ -0,0 +1,175 @@
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.netty.TieredStorageNettyService;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageMemoryManager;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageResourceRegistry;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.tier.TierProducerAgent;
+import org.apache.flink.util.ExceptionUtils;
+
+import java.io.IOException;
+import java.util.Arrays;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/** The memory tier implementation of {@link TierProducerAgent}. */
+public class MemoryTierProducerAgent implements TierProducerAgent, NettyServiceProducer {
+
+    private final int segmentSizeBytes;
+
+    private final int segmentSizeBuffers;
+
+    private final TieredStorageMemoryManager memoryManager;
+
+    private final boolean isBroadcastOnly;
+
+    // Record the byte number currently written to each subpartition.
+    private final int[] numSubpartitionEmitBytes;
+
+    /**
+     * Whether a subpartition has been registered to the netty service. The array index is
+     * corresponding to the subpartition id.
+     */
+    private final boolean[] nettyServiceRegistered;
+
+    private final MemoryTierSubpartitionProducerAgent[] subpartitionProducerAgents;
+
+    public MemoryTierProducerAgent(
+            TieredStoragePartitionId partitionId,
+            int numSubpartitions,
+            int bufferSize,
+            int segmentSizeBytes,
+            boolean isBroadcastOnly,
+            TieredStorageMemoryManager memoryManager,
+            TieredStorageNettyService nettyService,
+            TieredStorageResourceRegistry resourceRegistry) {
+        checkArgument(segmentSizeBytes >= bufferSize, "One segment contains at least one buffer.");
+
+        this.segmentSizeBytes = segmentSizeBytes;
+        this.segmentSizeBuffers = segmentSizeBytes / bufferSize;
+        this.isBroadcastOnly = isBroadcastOnly;
+        this.memoryManager = memoryManager;
+        this.numSubpartitionEmitBytes = new int[numSubpartitions];
+        this.nettyServiceRegistered = new boolean[numSubpartitions];
+        this.subpartitionProducerAgents = new MemoryTierSubpartitionProducerAgent[numSubpartitions];
+
+        Arrays.fill(numSubpartitionEmitBytes, 0);
+        nettyService.registerProducer(partitionId, this);
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; ++subpartitionId) {
+            subpartitionProducerAgents[subpartitionId] =
+                    new MemoryTierSubpartitionProducerAgent(subpartitionId, nettyService);
+        }
+        resourceRegistry.registerResource(partitionId, this::releaseResources);
+    }
+
+    @Override
+    public boolean tryStartNewSegment(TieredStorageSubpartitionId subpartitionId, int segmentId) {
+        if (isBroadcastOnly) {
+            return false;
+        }
```

Comment:
For broadcast only result partition, the memory tier shouldn't be created at all. So this should be a check-state, and it should be mored to the constructor.

Reply:
Fixed, replaced this with `checkState`.